dbms/src/AggregateFunctions/AggregateFunctionMinMaxWindow.h

```diff
@@ -92,74 +86,59 @@ struct SingleValueDataStringForWindow
 private:
     using Self = SingleValueDataStringForWindow;
 
-    struct StringWithCollator
-    {
-        StringWithCollator(const StringRef & value_, TiDB::TiDBCollatorPtr collator_)
-            : value(value_)
-            , collator(collator_)
-        {}
-
-        StringRef value;
-        TiDB::TiDBCollatorPtr collator;
-    };
+    mutable std::deque<StringRef> queue;
+    TiDB::TiDBCollatorPtr collator{};
 
-    struct Less
+public:
+    void insertResultInto(IColumn & to) const
     {
-        constexpr bool operator()(const StringWithCollator & left, const StringWithCollator & right) const
-        {
-            if unlikely (left.collator == nullptr)
-                return left.value < right.value;
-            return left.collator->compareFastPath(left.value.data, left.value.size, right.value.data, right.value.size);
-        }
-    };
+        if likely (!queue.empty())
+            static_cast<ColumnString &>(to).insertDataWithTerminatingZero(queue.front().data, queue.front().size);
+        else
+            static_cast<ColumnString &>(to).insertDefault();
+    }
 
-    using multiset = std::multiset<StringWithCollator, Less>;
+    void reset() { queue.clear(); }
 
-    mutable multiset saved_values;
-    TiDB::TiDBCollatorPtr collator{};
+    void decrease(const IColumn & column, size_t row_num)
+    {
+        assert(!queue.empty());
 
-    void saveValue(const StringRef & value) { saved_values.insert(StringWithCollator(value, collator)); }
+        auto str = static_cast<const ColumnString &>(column).getDataAtWithTerminatingZero(row_num);
+        if (collator != nullptr)
```

Comment:
I think here we don't need to use collator to compare, the value must be exactly the same if it can be removed.

Reply:
> I think here we don't need to use collator to compare, the value must be exactly the same if it can be removed.

done